docs/developer-guide/rules.md

```diff
@@ -1,57 +1,81 @@
-2. Once you have something to show, create a [pull request](https://github.com/stylelint/stylelint/compare).
+Some rules require extra flexibility to address a variety of use-cases. These can use a **secondary options object**. In `"color-hex-case": "upper"`, there is no secondary options object; in `"indentation": [2, { "except": ["block"] }]`, the secondary options object is `{ "except": ["block"] }`.
 
-## Options
+The most typical secondary options are `"ignores"` and `"except"`; but anything is possible.
 
-Use explicit, rather than implicit, options. For example:
+*Use explicit, rather than implicit, options.* For example:
 
 - `color-hex-case: "upper"|"lower"` rather than `color-hex-uppercase: "always"|"never"`
 
-As `color-hex-uppercase: "never"` *implies* always lowercase.
+`color-hex-uppercase: "never"` *implies* always lowercase, but `color-hex-case: "lower"` makes it *explicit*.
 
-## Messages
+### Determine warning messages
 
-Take the form of:
+Messages usually take one of these forms:
 
-- "Expected a ... something".
-- "Unexpected ... something" (for rejection e.g. when something is disallowed).
+- "Expected \[something\] \[in some context\]".
+- "Unexpected \[something\] \[in some context\]."
 
-## README
+Look at the messages of other rules to glean more conventions and patterns.
 
-Each rule must be accompanied by a README, which takes the form of:
+### Write the rule
 
-1. Rule name.
-2. Single line description.
-3. Code example (if necessary).
-4. Expanded description (if necessary).
-5. Options (if applicable).
-6. Example patterns that are considered warnings (for each option value).
-7. Example patterns that are *not* considered warnings (for each option value).
-8. Optional options (if applicable).
+*When writing the rule, always look to other similar rules for conventions and patterns to start from and mimic.*
 
-## Single line descriptions
+You will use the simple [PostCSS API](https://github.com/postcss/postcss/blob/master/docs/api.md) to navigate and analyze the CSS syntax tree.
 
-Take the form of:
+Depending on the rule, you may also want ot use [postcss-value-parser](https://github.com/TrySound/postcss-value-parser) and/or [postcss-selector-parser](https://github.com/postcss/postcss-selector-parser), which are easier and more accurate than most people's guesses at regular expressions.
 
-- "Disallow ..." (for `no` rules).
-- "Limit ..." (for `max` rules).
-- "Require ..." (for `after`, `before` and `inside` rules).
-- "Specify ..." (for everything else).
+stylelint has a number of [utility functions](https://github.com/stylelint/stylelint/tree/master/src/utils) that are used in existing rules and might prove useful to you, as well. Please look through those so that you know what's available. (And if you have a new function that you think might prove generally helpful, let's add it to the list!)
```

Comment:
Do we want to single out the `optionsValidator` here, as it's applicable to every rule?